x-pack/plugins/observability/public/pages/slo_edit/hooks/use_unregister_fields.ts

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { IndicatorType } from '@kbn/slo-schema';
+import { assertNever } from '@kbn/std';
+import { useEffect, useState } from 'react';
+import { useFormContext } from 'react-hook-form';
+import {
+  APM_AVAILABILITY_DEFAULT_VALUES,
+  APM_LATENCY_DEFAULT_VALUES,
+  CUSTOM_KQL_DEFAULT_VALUES,
+  CUSTOM_METRIC_DEFAULT_VALUES,
+  SLO_EDIT_FORM_DEFAULT_VALUES,
+} from '../constants';
+import { CreateSLOForm } from '../types';
+
+/**
+ * This hook handles the unregistration of inputs when selecting another SLI indicator.
+ * We could not use shouldUnregister on the controlled form fields because of a bug when submitting the form
+ * which was unmounting the components and therefore unregistering the associated values.
+ */
+export function useUnregisterFields({ isEditMode }: { isEditMode: boolean }) {
+  const { watch, unregister, reset, resetField } = useFormContext<CreateSLOForm>();
+  const [indicatorTypeState, setIndicatorTypeState] = useState<IndicatorType>(
+    watch('indicator.type')
+  );
+  const indicatorType = watch('indicator.type');
+
+  useEffect(() => {
+    if (indicatorType !== indicatorTypeState && !isEditMode) {
+      setIndicatorTypeState(indicatorType);
+      unregister('indicator.params');
+      switch (indicatorType) {
+        case 'sli.metric.custom':
```

Comment:
Nit: what do you think of exporting these indicator types as const and grouping them in a separate file?

Reply:
These values are also used in the `constants` file.

Reply:
We don't need to, this indicatorType are typed, so you won't be able to add an invalid one, or forget to add a new one (due to the assertNever at the end)